### DISCUSSION TITLE
Kotlin Multiplatform

### DISCUSSION THREAD

**Discussion by serpro69** [serpro69/kotlin-faker#235] [Ideas]
Discussion around adding mutiplatform support to kotlin-faker, which is done in #226

**Comment by tamimattafi** [comment]
@serpro69 Have you ever tried building the project for macOS aarch64?
I'm getting this error while syncing the project with JDK 11 or 17
```kotlin
Build file '/Users/tamimattafi/Projects/kmm/kotlin-faker/cli-bot/build.gradle.kts' line: 64

Failed to calculate the value of task ':cli-bot:compileJava' property 'javaCompiler'.
> No matching toolchains found for requested specification: {languageVersion=8, vendor=matching('Temurin'), implementation=vendor-specific} for MAC_OS on aarch64.
   > No locally installed toolchains match and toolchain download repositories have not been configured.
```

Line 64 of `cli-bot/build.gradle.kts` is
```kotlin
"Class-Path" to project.configurations.compileClasspath.get().joinToString(" ") { it.name },
```

Commenting this line doesn't hide the issue, it's still present but with line 143 (The end of the file)

Changing IDE's JDK to 1.8 throws another error
```kotlin
A problem occurred configuring project ':cli-bot'.
> Could not resolve all files for configuration ':cli-bot:classpath'.
   > Could not resolve org.graalvm.buildtools:native-gradle-plugin:0.10.1.
     Required by:
         project :cli-bot > org.graalvm.buildtools.native:org.graalvm.buildtools.native.gradle.plugin:0.10.1
      > No matching variant of org.graalvm.buildtools:native-gradle-plugin:0.10.1 was found. The consumer was configured to find a library for use during runtime, compatible with Java 8, packaged as a jar, and its dependencies declared externally, as well as attribute 'org.gradle.plugin.api-version' with value '8.4' but:
          - Variant 'apiElements' capability org.graalvm.buildtools:native-gradle-plugin:0.10.1 declares a library, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component for use during compile-time, compatible with Java 11 and the consumer needed a component for use during runtime, compatible with Java 8
              - Other compatible attribute:
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.4')
          - Variant 'javadocElements' capability org.graalvm.buildtools:native-gradle-plugin:0.10.1 declares a component for use during runtime, and its dependencies declared externally:
              - Incompatible because this component declares documentation and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them packaged as a jar)
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.4')
          - Variant 'runtimeElements' capability org.graalvm.buildtools:native-gradle-plugin:0.10.1 declares a library for use during runtime, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component, compatible with Java 11 and the consumer needed a component, compatible with Java 8
              - Other compatible attribute:
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.4')
          - Variant 'sourcesElements' capability org.graalvm.buildtools:native-gradle-plugin:0.10.1 declares a component for use during runtime, and its dependencies declared externally:
              - Incompatible because this component declares documentation and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them packaged as a jar)
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.4')
```
Searching for a solution.

**Reply by serpro69** [reply]
I don't have a Mac, so haven't tried to compile locally to that target. But are you trying to compile from master just for the JVM on a mac machine with the current (non-multiplatform) build setup, or have you updated the build to use kotlin-multiplatform plugin and trying to compile to a mac target?

Regardless, the jdk requirements when building the project are a bit ... weird, to say the least. It's described in the [contributing#prerequisites](https://github.com/serpro69/kotlin-faker/blob/master/CONTRIBUTING.adoc#prerequisites) and basically you need to have 3 jdk versions installed on the machine to build the entire project.

But can you try to exclude the cli module altogether from the build process? Would that build OK? In that case you would only need to jdk8 to build everything else.
If that works, I think it should be OK to just exclude it entirely, and then fix it later.

**Reply by serpro69** [reply]
Here's a git diff for excluding the cli module altogether. I've just tested it myself and I can build everything else with just jdk8.

```diff
diff --git a/bom/build.gradle.kts b/bom/build.gradle.kts
index cde7bffe..086d5c12 100644
--- a/bom/build.gradle.kts
+++ b/bom/build.gradle.kts
@@ -26,7 +26,7 @@ val isRelease = provider {
 
 // Exclude subprojects that will never be published so that when configuring this project
 // we don't force their configuration and do unnecessary work
-val excludeFromBom = listOf(":cli-bot", ":docs", ":extension", ":faker", ":test", ":extension:kotest-property-test")
+val excludeFromBom = listOf(/*":cli-bot", */":docs", ":extension", ":faker", ":test", ":extension:kotest-property-test")
 fun projectsFilter(candidateProject: Project) =
     excludeFromBom.none { candidateProject.path == it }
         && candidateProject.name != bom.name
diff --git a/build.gradle.kts b/build.gradle.kts
index 07e444ba..52e20c3d 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -172,5 +172,5 @@ tasks.withType<TagTask>().configureEach {
 }
 
 apiValidation {
-    ignoredProjects += listOf("bom", "cli-bot", "docs", "test")
+    ignoredProjects += listOf("bom", /*"cli-bot", */"docs", "test")
 }
diff --git a/settings.gradle.kts b/settings.gradle.kts
index e8b4fb2b..0b8c676d 100644
--- a/settings.gradle.kts
+++ b/settings.gradle.kts
@@ -15,7 +15,7 @@ rootProject.name = "kotlin-faker"
 include(
     ":bom",
     ":core",
-    ":cli-bot",
+    // ":cli-bot",
     ":docs",
 )
 
```

Try if this works for you, and if so, then I think we can just keep the cli ignored for now.

**Reply by tamimattafi** [reply]
@serpro69 
Yes, I tried to compile from master without any multiplatform setup. I tried to exclude the `cli-bot` from `settings.gradle.kts` but it seems like that wasn't enough, I will try it with your instruction.

As for JDK versions, I will take a look on the prerequisites. 
I will keep this thread updated.

**Reply by serpro69** [reply]
In short, the problem with `cli` module is that it uses a graal gradle plugin that seems to require jdk version >= 11 (there's no mention of that in the docs that I could see, but I couldn't get it to work with java8)
Graal-CE distribution itself starts from version 17. But the gradle process needs to be running with java 11 because that's maximum supported version for a plugin in the `docs` module. So it's a small mess because of these hard dependencies on different java versions....

I'm now starting to look into refactoring docs and getting rid of the old unsupported framework that is currently used to build documentation. But that will only solve half of the problem and CLI would still need to be tackled sooner or later. I'm not sure yet what to do with it, so I'll have to take a look into it once I have some time. But since it's not part of the library itself, and isn't that critical, I think it should be OK to just ignore it for now so it doesn't become a blocker. If need be, I can just keep compiling it locally for awhile, in worst case.

What version of java do you have locally btw? And what vendor? I think this could also be causing some issues if you have a different vendor locally: https://github.com/serpro69/kotlin-faker/blob/d9cd70a5482df554ef2e534203e1a9188344e5fd/cli-bot/build.gradle.kts#L54
(I've just removed this part, so it should't be tied to a specific vendor anymore. Have no idea why I added that... and seems like that's the only place where vendor was specified explicitly)

But anyways, just try to exclude cli completely, and if that works - let's keep it excluded for now.

**Reply by tamimattafi** [reply]
@serpro69 Having JDK installed as mentioned in the prerequisites fixed the problem, and even tho there's an issue on my machine with graalvm-17, the sync completed successfully. Thank you for the tips.

I will study the build config and start adding the multiplatform plugin this week, hopefully on the weekend I will have something to show.

**Reply by serpro69** [reply]
Awesome! Feel free to post more questions here if/when you need further help.

**Comment by tamimattafi** [comment]
@serpro69 Hello!
Unfortunately, I didn't make any significant progress this week due to tight time at work. However, I studied the build config and made some observations, which might be important for further steps.

1. The current config uses `maven-publish` to upload artifacts to `mavenCentral`, however, this might not work seamlessly for Kotlin Multiplatform projects. I suggest using [com.vanniktech.maven.publish](https://github.com/vanniktech/gradle-maven-publish-plugin) plugin since it handles all artifact types. Here's an example of how to set it up: [here]( https://github.com/tamimattafi/krop/blob/feat/multiplatform/plugins/publishing/src/main/kotlin/com/mr0xf00/easycrop/publish/PublishConventions.kt)
2. It might be worth it to move from `buildSrc` to `includeBuild` and break plugins into separate modules, for example: [here](https://github.com/tamimattafi/krop/tree/feat/multiplatform/plugins). This will make build logic cleaner and more flexible
3. Not sure about the use-case differences between `faker-lib` and `faker-ext` conventions, what is each of them used for?
4. What is `faker-provider` conventions used for?

The first step will require implementing kotlin multiplatform plugin with basic targets, and move all the current logic from the `main` source set to `jvmMain`, and `test` to `jvmTest` (Basically rename the directories). This will lay the foundation for creating new source sets such as `commonMain` and move shared logic eventually. However, this should be done and merged quickly to avoid potential conflicts.

**Reply by serpro69** [reply]
Hi @tamimattafi ,
1. I think `maven-publish` should support multiplatform, but haven't explored it too much. But e.g. libs like kotlin-coroutines use maven publish plugin as well - https://github.com/Kotlin/kotlinx.coroutines/blob/cd696d3f8f4afdbc735915c2bded974616331b55/buildSrc/src/main/kotlin/pub-conventions.gradle.kts#L8
2. Yeah, buildSrc has some duplications and isn't very well structured. I'm in the process of refactoring that. Not sure about separate modules for plugins, that may be an overkill in most cases, but I want to extract some common logic as you've probably seen it's duplicated in several *convention plugins.
3. `faker-lib` plugin is the "base" plugin whereas `faker-provider` has extra stuff for [faker](https://github.com/serpro69/kotlin-faker/tree/master/faker) submodules; `faker-ext` is for the [extension](https://github.com/serpro69/kotlin-faker/tree/master/extension) submodules

**Reply by tamimattafi** [reply]
@serpro69 
1. Indeed `maven-publish` does support  all types of publications, since it's the base plug-in of that one I've sent. The case is configuring it might require extra time depending on your dependencies and targets, which is done out of the box by the second plug-in, and it won't require any changes if you add more targets or change your build script. Also, it already supports the new Central Portal, since s01 will no longer accept new namespaces, and might require migration for old ones in the future. The plug-in does this switch in one line and complies with all requirements of each registry.
2. What I meant is a separate module or at least folder for each type of plug-in, there should be separate plug-ins for multiplaform conventions and for publishing conventions. They might have base logic.
3. Understood

**Reply by serpro69** [reply]
> Indeed maven-publish does support all types of publications, since it's the base plug-in of that one I've sent. The case is configuring it might require extra time depending on your dependencies and targets, which is done out of the box by the second plug-in, and it won't require any changes if you add more targets or change your build script. Also, it already supports the new Central Portal, since s01 will no longer accept new namespaces, and might require migration for old ones in the future. The plug-in does this switch in one line and complies with all requirements of each registry.

Yeah, could make sense if it reduces some complexity. I'm not against it :) 

> What I meant is a separate module or at least folder for each type of plug-in, there should be separate plug-ins for multiplaform conventions and for publishing conventions. They might have base logic.

But that's what convention plugins in buildsrc are supposed to do, right? You create simple convention plugins like `publish-conventions.gradle.kts` and `mp-conventions.gradle.kts`, each handling their own part of the build process, e.g. publishing and multiplatform.

**Reply by tamimattafi** [reply]
> But that's what convention plugins in buildsrc are supposed to do, right? You create simple convention plugins like publish-conventions.gradle.kts and mp-conventions.gradle.kts, each handling their own part of the build process, e.g. publishing and multiplatform.

Exactly, so if you are not against such separation, we could have different plugins for different parts and apply them both or separately to certain modules.

I will then start with writing these plugins in `buildSrc`, and after your review, we will apply them to the modules.

**Reply by serpro69** [reply]
> I will then start with writing these plugins in buildSrc, and after your review, we will apply them to the modules.

I'm actually doing that already as I mentioned above :) I'm just missing a few final bits before it's finished :)

**Reply by serpro69** [reply]
Made a PR for the above - https://github.com/serpro69/kotlin-faker/pull/241 
I'm still testing it, but feel free to provide some comments based on your experience if you'd like :)

**Reply by tamimattafi** [reply]
@serpro69 Seems good to me, now it will be easier to add the multiplatform plugin and adjust publishing scripts :)

**Reply by tamimattafi** [reply]
@serpro69 I see you've already merged #241, are those the final changes? I'm asking, due to [this](https://github.com/serpro69/kotlin-faker/actions/runs/9284249209/job/25546215134) failure, should I sync my fork with your repository or not yet?

**Reply by serpro69** [reply]
@tamimattafi , that's looks like a flaky test. Testing random data generation is.... tricky, to say the least, and some tests fail now and then. Unfortunately I haven't found a way to automatically restart a failing test with kotest, not sure if that's possible at all.
I'm re-running the job now, I think it should go back to green since it passed in the PR before I merged.

UPD: back in the :green_circle:

**Reply by tamimattafi** [reply]
@serpro69 It seems like implementing the multiplatform plugin is not as easy as I thought, I tried doing simple changes, and I found my self constantly breaking stuff in other build scripts and plugins 😂

Maybe you will have better luck since you know the config better? At the moment, all you need is to do these changes while keeping everything working as supposed:
1. Change `kotlin("jvm")` to `kotlin("multiplatform")` basically everywhere
2. The `java { }` block will be available since the multiplatform plugin provides `JavaPluginExtension`, therefore you might not need `java-library` anymore
3. Add the multiplatform configuration with `jvm` target only (For now)
```kotlin
plugins {
    kotlin("multiplatform")
}

kotlin {
    applyDefaultHierarchyTemplate()
    jvmToolchain(8) // Mabye consider using a newer version
    
    // Targets
    jvm()
}
```
4. If you have `src/main/java` (I only found this in the `:core` module) then you might consider renaming it to `src/main/kotlin` since it won't be available as a source set with the new plugin (Maybe you can add it manually using `srcDir`)
5. Your `main` source set will work just fine since we only have a `jvm` target, but ideally you should rename it to `jvmMain` and use the new dependencies block as shown below:
```kotlin
plugins {
    id("faker-lib-conventions")
}

kotlin {
    sourceSets {
        jvmMain.dependencies { 
            implementation("...")
            api("...")
        }
     }
}
```
6. As in the previous point, the same is correct for the `test` source set, but ideally it should become `jvmTest` and use the new dependencies block `jvmTest.dependencies`. But If you need more control over a source set, you can do something like this:
```kotlin
sourceSets {
    val jvmTest by getting {
        dependsOn(jvmMain)
        dependencies {
            // ...
        }
    }
}
```

That's basically it, after that you should consider creating `commonMain` source set and start moving pure Kotlin logic step by step, and later on, you might add more targets and use `expect / actual` for platform specific logic.

I recently moved a simple library to multiplatform, which can be taken as an example. We might have different approaches when it comes to build logic, but I'm sure you will find the diff useful. Here's the pull request (You can check the second commit only since it has the most important changes) https://github.com/mr0xf00/easycrop/pull/12

As for publishing, in this step you probably won't need to change anything since having only `jvm` target doesn't change publication artifacts. But if you have spare time, consider moving to the [plugin](https://github.com/vanniktech/gradle-maven-publish-plugin) that I mentioned earlier, especially if you are considering migration to the new Central Portal with minimal efforts.

I would be happy to answer your questions if you have any

**Reply by serpro69** [reply]
Thanks for the inputs and details @tamimattafi . Yeah, I can definitely try :)  
In fact, I've started doing something similar in another lib which has a simpler setup, just to learn multiplatform stuff - https://github.com/serpro69/semver.kt/issues/31 .

**Reply by tamimattafi** [reply]
@serpro69, I can take a look at the Pull Request and give some thoughts when you are done, if you don't mind :D

**Reply by serpro69** [reply]
@tamimattafi not at all, thanks! I'll post it here or mention you in the PR once it's ready, but don't expect anything very soon - a bit busy at work at the moment :D

**Reply by tamimattafi** [reply]
> a bit busy at work at the moment :D 

I couldn't relate more 🥲

**Reply by serpro69** [reply]
Hey @tamimattafi :wave: ,
...I've been away from this project for a <s>little</s> :see_no_evil: while and only had time for maintenance and small improvements the past year or so, unfortunately. But I spent some time today on this and I think I managed to convert the build configuration to use kotlin multiplatform plugin (as in, tests pass fine via `./gradlew clean jvmTest`; need to check other stuff also like publishing and so on).
I had to make some semi-drastic changes, particularly removing 'shadow' plugin completely (I can't get it to work yet, and atm can't be bothered to do so either :grin: ), and I've also merged integration tests with unit tests to simplify the whole gradle code.
There's a draft PR here if you'd like to take a look: https://github.com/serpro69/kotlin-faker/pull/278

**Reply by tamimattafi** [reply]
Hey @serpro69!
Great news! I took a quick look, everything looks good to me. 

As for publishing, that's normal that a separate component will be created for each sourceSet/platform, for example, `module` for common, `module-jvm` for JVM, `module-mingwx64` and so on. I believe the Kotlin team is making (Or already made) good progress with klib to generate fewer artifacts for publication, but I haven't checked it in a long time.
I suggest using a publishing plugin, such as [com.vanniktech.maven.publish](https://github.com/vanniktech/gradle-maven-publish-plugin), to avoid some headaches. Here's an [example](https://github.com/tamimattafi/krop/blob/main/plugins/publishing/src/main/kotlin/com/attafitamim/krop/publish/PublishConventions.kt) of a convention plugin.

As for the Gradle scripts, everything looks neat. I can suggest doing a minor change to make code more concise, but it's more of a taste than a rule, and it works for some source sets only, not all:
```kts
kotlin {
     sourceSets {
          // instead of this
          val androidMain by getting {
              dependencies {
                   // add something here
              }
          }
          
          // you can do this
          androidMain.dependencies {
              // add something here
          }
     }
}
```

Great work! 
I hope you will soon be able to support more platforms in addition to JVM!

**Reply by serpro69** [reply]
Just a small update: had a bit more time to spend on this, and I think I mostly fixed the publishing logic. Want to test it a bit more and clean up, but at least all artifacts pass validation in central now.

**Reply by serpro69** [reply]
If you want to test, I've now published the snapshot versions of the artifacts.

```kotlin
repositories {
    mavenCentral()
    maven {
        url = URI("https://central.sonatype.com/repository/maven-snapshots/")
    }
}

dependencies {
    testImplementation(kotlin("test"))
    testImplementation(platform("io.github.serpro69:kotlin-faker-bom:2.1.0-SNAPSHOT"))
    testImplementation("io.github.serpro69:kotlin-faker-jvm") // can also use dependency w/o -jvm qualifier
    testImplementation("io.github.serpro69:kotlin-faker-ext-blns-jvm")
    testImplementation("io.github.serpro69:kotlin-faker-books-jvm")
}
```
